### PR TITLE
[CARBONDATA-2999] support read schema from S3

### PR DESCRIPTION
**Pull request by xubo245**
[CARBONDATA-2999] support read schema from S3
SDK and CSDK both support it.

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [x] Any interfaces changed?
 add
 - [x] Any backward compatibility impacted?
 No
 - [x] Document update required?
Yes
 - [x] Testing done
      add test case
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
JIRA2951

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1457/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9715/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1667/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1458/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1668/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9716/

**Comment by xubo245** [comment]
@KanakaKumar @BJangir @ajantha-bhat @jackylk Please review it.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1501/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1711/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9759/

**Comment by xubo245** [comment]
@KanakaKumar @jackylk @QiangCai @ajantha-bhat @kunal642 Rebased, Please review it.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1614/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/9874/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/1825/

**Comment by xubo245** [comment]
@KanakaKumar @jackylk @QiangCai @ajantha-bhat @kunal642 Rebased, Please review it.

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonSchemaReader.java [hunk 1]
```diff
@@ -147,34 +170,55 @@ public static Schema readSchema(String path, boolean validateSchema) throws IOEx
...
    * @param dataFilePath carbondata file store path
+   * @param conf         hadoop configuration support, can set s3a AK,SK,
+   *                     end point and other conf with this
    * @return Schema object
    * @throws IOException
    */
   @Deprecated
-  public static Schema readSchemaInDataFile(String dataFilePath) throws IOException {
-    return readSchema(dataFilePath, false);
+  public static Schema readSchemaInDataFile(String dataFilePath, Configuration conf)
```
Comment (KanakaKumar) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonSchemaReader.java]: No need to change the deprecated old method.  Keep as it is.

Comment (xubo245) [reply store/sdk/src/main/java/org/apache/carbondata/sdk/file/CarbonSchemaReader.java]: ok, done

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2021/

**Comment by xubo245** [comment]
rebase and fix the conflict

**Comment by KanakaKumar** [comment]
LGTM

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1816/

File: store/CSDK/test/main.cpp [hunk 2]
```diff
@@ -822,16 +831,17 @@ int main(int argc, char *argv[]) {
     // init jvm
     JNIEnv *env;
     env = initJVM();
-    char *S3WritePath = "s3a://csdk/WriterOutput/carbondata2";
-    char *S3ReadPath = "s3a://csdk/WriterOutput/carbondata";
+    char *S3WritePath = "s3a://xubo/WriterOutput/carbondata2";
```
Comment (QiangCai) [inline store/CSDK/test/main.cpp]: please change the path

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10075/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/1819/

File: store/CSDK/test/main.cpp (hunk 2 above)
Comment (xubo245) [reply store/CSDK/test/main.cpp]: OK，done

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2028/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10077/

**Comment by QiangCai** [comment]
LGTM

**Pull request closed without merging**